src/timer.rs

```diff
@@ -146,125 +179,149 @@ macro_rules! timer {
+                    let aclk_ratio: u32;
+                    let real_prescaler : ClockSourcePrescaler;
+                    let smclk_ratio : u32;
+                    let tick_count : u16;
+                    let mut count_ratio : u32;
+                    let mut clock_source : ClockSource;
+                    let mut is_config: bool;
+                    let mut min_prescaler : u8;
+
+                    is_config = false;
+                    count_ratio = 0;
+                    clock_source = Smclk;
+
+                    if count.1 == Hertz {
+                        let frequency = count.0;
+                        aclk_ratio = self.clocks.aclk.0 / frequency;
+                        smclk_ratio = self.clocks.smclk.0 / frequency;
                     } else {
-                        false
+                        let period = count.0;
+                        aclk_ratio = (period*self.clocks.aclk.0)/1000;
+                        smclk_ratio = period*(self.clocks.smclk.0/1000);
                     }
-                }
 
-                fn setup_frequency(&self, frequency: u32) -> bool {
-                    let max_period = MAX_PRESCALER * MAX_COUNT;
+                    if(smclk_ratio < max_period) {
+                        count_ratio = smclk_ratio;
+                        clock_source = Smclk;
+                    } else if (aclk_ratio < max_period) {
+                        count_ratio = aclk_ratio;
+                        clock_source = Aclk;
+                    }
 
-                    if (frequency > (self.clocks.smclk.0 / max_period)) {
-                        let prescaler = self.get_prescaler((self.clocks.smclk.0/(MAX_COUNT * frequency)) as u8);
-                        let count = (self.clocks.smclk.0 / (frequency * prescaler[2] as u32)) as u16;
-                        self.set_clock_source(0x08, prescaler[1]);
-                        self.setup_count(count);
-                        true
-                    } else if (frequency > (self.clocks.aclk.0 / max_period)) {
-                        let prescaler = self.get_prescaler((self.clocks.aclk.0/(MAX_COUNT * frequency)) as u8);
-                        let count = (self.clocks.aclk.0 / (frequency * prescaler[2] as u32)) as u16;
-                        self.set_clock_source(0x04, prescaler[1]);
-                        self.setup_count(count);
-                        true
-                    } else {
-                        false
+                    if count_ratio != 0 {
+                        min_prescaler = (count_ratio / MAX_COUNT) as u8;
+
+                        // In period, prescaler need to be above the min_prescaler value
+                        if count.1 != Hertz {
+                            min_prescaler = min_prescaler + 1;
+                        }
+
+                        real_prescaler = self.get_prescaler(min_prescaler);
+                        tick_count = (count_ratio / real_prescaler as u32) as u16;
```

Comment:
There's no need to declare these vals before actually needing them

```suggestion
                        let real_prescaler = self.get_prescaler(min_prescaler);
                        let tick_count = (count_ratio / real_prescaler as u32) as u16;
```